deep_quoridor/src/agents/alphazero/multiprocess_evaluator.py

```diff
@@ -0,0 +1,223 @@
+        if game.is_rotated():
+            policy = policy[self._action_mapping_rotated_to_original]
+
+        return value, policy
+
+    def evaluate_arrays(self, input_arrays: np.ndarray, action_mask_arrays: np.ndarray):
+        self._request_queue.put((input_arrays, action_mask_arrays, self._client_id))
+        values_array, policies_array = self._result_queue.get()
+        return values_array, policies_array
+
+    def rotate_policy_from_original(self, policy: np.ndarray):
+        return policy[self._action_mapping_original_to_rotated]
+
+    def rotate_if_needed_to_point_downwards(self, game: Quoridor):
+        return NNEvaluator.rotate_if_needed_to_point_downwards(game)
+
+    def game_to_input_array(self, game):
+        return NNEvaluator.game_to_input_array(game)
+
+    def clear_cache(self):
+        self._cache.clear()
+
+    def get_statistics(self):
+        return compute_statistics(self._evaluation_log)
+
+
+class EvaluatorServer(threading.Thread):
+    def __init__(
+        self,
+        evaluator: NNEvaluator,
+        input_queue: mp.Queue,
+        output_queues: list[mp.Queue],
+        max_log_len=int(1e8),
+    ):
+        super().__init__()
+        self._evaluator = evaluator
+        self._cache = {}
+        self._input_queue = input_queue
+        self._output_queues = output_queues
+        self._max_l = max_log_len
+        self._evaluation_log = deque(maxlen=max_log_len)
+        self._shutdown = False
+
+    def run(self):
+        while not self._shutdown:
+            try:
+                inputs_array, action_masks_array, client_id = self._input_queue.get(timeout=0.1)
+            except queue.Empty:
+                continue
+
+            values_array, policies_array = self._evaluate_arrays(inputs_array, action_masks_array)
+            self._send_result_to_client(client_id, values_array, policies_array)
+
+            # Update the cache
+            for row_i in range(len(values_array)):
+                self._cache[inputs_array[row_i].tobytes()] = (values_array[row_i], policies_array[row_i])
+
+    def _evaluate_arrays(self, inputs_array, action_masks_array):
+        evaluation_start_time = time.time()
+        used_cache = False
+
+        # If the first input array is in our cache, we return the cached value, policy for that
+        # one and don't evaluate the others.
+        cache_key = inputs_array[0].tobytes()
+        if cache_key in self._cache:
+            value, policy = self._cache.get(cache_key)
+            values_array = np.expand_dims(value, 0)
+            policies_array = np.expand_dims(policy, 0)
+            used_cache = True
+        else:
+            inputs_tensor = torch.from_numpy(inputs_array).to(my_device())
+            action_masks_tensor = torch.from_numpy(action_masks_array).to(my_device())
+
+            # Do the actual evalution
+            values_tensor, policies_tensor = self._evaluator.evaluate_tensors(inputs_tensor, action_masks_tensor)
```

Comment:
I realized something: we are evaluating all the "extra games", but some of them may be in the cache already, because they could have been used in a previous game or even in this same game.  Not sure if this is worth trying, but we could remove from the input everything that is already in the cache (can be done separately).